src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/HPackDecoder.cs

```diff
@@ -124,107 +122,270 @@ public void Decode(in ReadOnlySequence<byte> data, bool endHeaders, IHttpHeaders
+                    // 0 should always be represented by a single byte, so we shouldn't need to check for it in the continuation case.
+                    Debug.Assert(intResult != 0, "A header name length of 0 should never be encoded with a continuation byte.");
+
+                    OnStringLength(intResult, nextState: State.HeaderName);
+                    currentIndex++;
+                    ParseHeaderName(data, ref currentIndex, handler);
+                    break;
+                }
+            }
+        }
+
+        private void ParseHeaderValueLengthContinue(ReadOnlySpan<byte> data, ref int currentIndex, IHttpHeadersHandler? handler)
+        {
+            for (; currentIndex < data.Length; currentIndex++)
+            {
+                if (_integerDecoder.TryDecode(data[currentIndex], out var intResult))
+                {
+                    // IntegerDecoder disallows overlong encodings where an integer is encoded with more bytes than is strictly required.
+                    // 0 should always be represented by a single byte, so we shouldn't need to check for it in the continuation case.
+                    Debug.Assert(intResult != 0, "A header value length of 0 should never be encoded with a continuation byte.");
+
+                    OnStringLength(intResult, nextState: State.HeaderValue);
+                    currentIndex++;
+                    ParseHeaderValue(data, ref currentIndex, handler);
+                    break;
+                }
+            }
+        }
+
+        private void ParseHeaderFieldIndex(ReadOnlySpan<byte> data, ref int currentIndex, IHttpHeadersHandler? handler)
+        {
+            for (; currentIndex < data.Length; currentIndex++)
+            {
+                if (_integerDecoder.TryDecode(data[currentIndex], out var intResult))
+                {
+                    OnIndexedHeaderField(intResult, handler);
+                    currentIndex++;
+                    break;
+                }
+            }
+        }
+
+        private void ParseHeaderNameIndex(ReadOnlySpan<byte> data, ref int currentIndex, IHttpHeadersHandler? handler)
+        {
+            for (; currentIndex < data.Length; currentIndex++)
+            {
+                if (_integerDecoder.TryDecode(data[currentIndex], out var intResult))
+                {
+                    OnIndexedHeaderName(intResult);
+                    currentIndex++;
+                    ParseHeaderValueLength(data, ref currentIndex, handler);
+                    break;
+                }
+            }
+        }
+
+        private void ParseHeaderNameLength(ReadOnlySpan<byte> data, ref int currentIndex, IHttpHeadersHandler? handler)
+        {
+            if (currentIndex < data.Length)
+            {
+                var b = data[currentIndex++];
+
+                _huffman = (b & HuffmanMask) != 0;
```

Comment:
Huffman check is done both here and ParseHeaderValueLength. Should there be a helper like ParseStringLength that's used in both places?